### ISSUE TITLE
Feature Request: surface user code TalonScript errors more aggressively and with more context

### ISSUE THREAD

**Issue by rntz** [talonvoice/beta#127]
Errors in user python code get reported quite helpfully: they appear in the talon log marked with ERROR and a trace back, and also cause a notification. This makes it very easy to notice that you've screwed up your python code. Errors in TalonScript user code are harder to notice. I'm thinking in particular of parse errors and uses of incorrectly named lists and captures. These do show up in the talon log, where they look like this:

```
2020-09-28 22:05:52    IO Rule not found: user.symbol
2020-09-28 22:05:52    IO List not found: user.code_libraries
2020-09-28 22:07:05    IO ParseError: did not understand definition on line 1: 'fubar ba'
```

But these lines are marked with IO, so they aren't visually distinct from more innocuous lines, and they do not cause notifications, which makes it harder to notice them; and [*] they don't come with a trace back or other indication of where the error occurred, which makes it harder to debug them.

It would be nice if these were also marked with ERROR or WARNING, or at least something visually distinct from IO, and if the "rule not found" and "list not found" errors indicated the code responsible in some way.

[*] Except for the ParseError, which comes with a line number - you can also infer the file name from the DEBUG line above it indicating the file that was loaded. So that's all right.

**Comment by rntz** [comment]
### Motivating code example for "Rule not found" error

Suppose that `keys.py` previously defined captures named `<user.symbol>` and `<user.number>` but they've been recently renamed `<user.symbol_key>`, `<user.number_key>`. However, formatters.py has a line that wasn't properly updated:

**formatters.py**
```
# NB. uses <user.symbol>, <user.number>
@ctx.capture(rule="(<user.symbol> | <user.number>)")
def formatter_immune(m): ...omitted...
```

This will produce the following at various points (it seems to happen whenever the DFA rules get built? and I'm not sure why user.number doesn't also get a rule not found error):

```
2020-09-28 22:55:38    IO COMPILING
2020-09-28 22:55:38    IO dfa rules built in 0.069s
2020-09-28 22:55:38    IO Rule not found: user.symbol
```

If this error message is overlooked, and if `<formatter_immune>` is used in a place where arbitrary text could be used instead (eg. see example below), the result is annoying to debug, because exactly the same words will be matched but the behavior will be different than expected.
```
@ctx.capture(rule="<self.formatters> <user.text> (<user.text> | <user.formatter_immune>)*")
def format_text(m): ...omitted...
```

I've used this example because it's how this problem actually came up, if you'd prefer a minimized example with no parts omitted I can make one.

**Comment by rntz** [comment]
I'm still trying to track down what's causing this "List not found: user.code_libraries" error, as it looks like user.code_libraries is actually properly declared. But in principle at least the same kind of thing could happen with lists instead of captures.

**Comment by lunixbochs** [comment]
Is it declared, but no context has initialized it to at least a blank list?

**Comment by rntz** [comment]
Indeed. Is the right thing to do to define it to be an empty list in an empty context?

**Comment by lunixbochs** [comment]
Yes, the list doesn't exist in the speech engine unless the value has been set somewhere.

**Comment by lunixbochs** [comment]
I think "define it to be an empty list in an empty context" won't be necessary in the next beta. I made an attempt to make module declared lists default to "empty list".
